### PR TITLE
Particles regression test

### PR DESCRIPTION
**Pull request by pgrete**
## PR Summary

Will eventually contain a regression test for the particle framework with MPI

## PR Checklist

- [ ] Code passes cpplint
- [ ] New features are documented.
- [ ] Adds a test for any bugs fixed. Adds tests for new features.
- [ ] Code is formatted
- [ ] Changes are summarized in CHANGELOG.md
- [ ] (@lanl.gov employees) Update copyright on changed files

### CODE REVIEW

**Comment by pgrete** [comment]
Follow up my review from last Friday I patched together sth that goes in the direction of what I had in mind in terms of a regression test for the particle framework.
In particular,
- creates a fixed amount of particles at the beginning using the `ProblemGenerator` function and never destroys them
- it uses leapfrog integration (currently with acceleration fields set to zero so that the particles move along straight lines) with `dt` determined by the particles velocities (and only makes one drift-kick-drift step per integration step)
- include an ugly output that I used for debugging

Using this simple test I noticed a few things where I'm not sure if I implemented/used the pieces of the Particle framework correctly or if there's a bug.
Thus, it'd be great to some additional pairs of eyes on this.

### Boundary conditions

Specifically, when I run the test until t=1 (which means all particles should be back at their original positions) using the default input file `./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog`
then
```
0 , 0 , 0 , 0.9 , 0.2 , 0.3 , 1 , 0 , 0
0 , 0 , 1 , 0.4 , 0.9 , 0.3 , 0 , 1 , 0
0 , 0 , 2 , -0.1 , 0.3 , 1.2 , 0 , 0 , 1
0 , 0 , 3 , 1.12 , 1.2 , 0.7 , 1 , 1 , 1

cycle=93 time=1.0000000000000000e+00 dt=8.1415712957866315e-03 zone-cycles/wsec_step=1.42e+08 wsec_step=2.88e-05

Driver completed.
time=1.00e+00 cycle=93
tlim=1.00e+00 nlim=100000
```
where the debug output is `rank, block_id, particle_id, x, y, z, vx, vy, vz`.
The `particle_id` is currently hardcoded to the index in the original initial condition array.
The original particle data is (with x,y,z,vx,vy,vz):
```
const std::array<std::array<Real, 6>, num_test_particles> particles_ic = {{
    {-0.1, 0.2, 0.3, 1.0, 0.0, 0.0},   // along x direction
    {0.4, -0.1, 0.3, 0.0, 1.0, 0.0},  // along y direction
    {-0.1, 0.3, 0.2, 0.0, 0.0, 1.0},  // along z direction
    {0.12, 0.2, -0.3, 1.0, 1.0, 1.0}, // along diagonal
}}
```
so it looks like (the hard coded IIRC) periodic boundary conditions are currently not working as I expect.

### Communication across MeshBlocks

Splitting the x dim in 2 meshblocks `./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog parthenon/meshblock/nx1=8` it looks like the particles never moved across blocks, e.g.,
```
cycle=8 time=8.6602540378443865e-02 dt=1.0825317547305485e-02 zone-cycles/wsec_step=7.53e+07 wsec_step=5.44e-05
0 , 0 , 0 , -0.00257214 , 0.2 , 0.3 , 1 , 0 , 0
0 , 0 , 2 , -0.1 , 0.3 , 0.297428 , 0 , 0 , 1
0 , 1 , 1 , 0.4 , -0.00257214 , 0.3 , 0 , 1 , 0
0 , 1 , 3 , 0.217428 , 0.297428 , -0.202572 , 1 , 1 , 1
cycle=9 time=9.7427857925749345e-02 dt=1.0825317547305485e-02 zone-cycles/wsec_step=6.98e+07 wsec_step=5.87e-05
0 , 0 , 0 , 0.00825318 , 0.2 , 0.3 , 1 , 0 , 0
0 , 0 , 2 , -0.1 , 0.3 , 0.308253 , 0 , 0 , 1
0 , 1 , 1 , 0.4 , 0.00825318 , 0.3 , 0 , 1 , 0
0 , 1 , 3 , 0.228253 , 0.308253 , -0.191747 , 1 , 1 , 1
```
particle 0 (third column) goes from `x < 0` to `x > 1` (fourth column) but stays on the same block (second column).
Same is true when running with MPI.
Did I miss a call to transfer particles between blocks? (which would also explain why the periodic boundary conditions didn't work).

Once we know what's going on I imagine to extend this example by a second particle type that does orbital advection, which would extend test coverage to a multiple species scenario.

File: example/particle_leapfrog/particle_leapfrog.cpp [hunk 1]
```diff
@@ -0,0 +1,455 @@
...
+  pkg->AddSwarmValue("vx", swarm_name, real_swarmvalue_metadata);
+  pkg->AddSwarmValue("vy", swarm_name, real_swarmvalue_metadata);
+  pkg->AddSwarmValue("vz", swarm_name, real_swarmvalue_metadata);
+
+  pkg->EstimateTimestepBlock = EstimateTimestepBlock;
+
+  return pkg;
+}
+
+AmrTag CheckRefinement(MeshBlockData<Real> *rc) { return AmrTag::same; }
```
Comment (Yurlungur) [inline example/particle_leapfrog/particle_leapfrog.cpp]: Is this actually needed? I think this should be the default, no?

File: example/particle_leapfrog/particle_leapfrog.cpp [hunk 2]
```diff
@@ -0,0 +1,455 @@
...
+}
+
+// initial particle position: x,y,z,vx,vy,vz
+constexpr int num_test_particles = 4;
+const std::array<std::array<Real, 6>, num_test_particles> particles_ic = {{
+    {-0.1, 0.2, 0.3, 1.0, 0.0, 0.0},   // along x direction
+    {0.4, -0.1, 0.3, 0.0, 1.0, 0.0},  // along y direction
+    {-0.1, 0.3, 0.2, 0.0, 0.0, 1.0},  // along z direction
+    {0.12, 0.2, -0.3, 1.0, 1.0, 1.0}, // along diagonal
+}};
```
Comment (Yurlungur) [inline example/particle_leapfrog/particle_leapfrog.cpp]: This strikes me as very manual. Perhaps that's fine. But it might be worth thinking about if this could be automated.

**Review by Yurlungur — COMMENTED** [review]
This is a nice suggestion. One thing it highlights for me is many of the tasking bits that are in the particles example should probably move to lower level... as utility functions in Parthenon itself if possible.

**Comment by brryan** [comment]
@pgrete I still need to dig into your test in detail but in the existing particles example it looks like particles are crossing at least some boundaries and meshblocks. Here's the output I get when I run a 1D problem with meshblocks that have `[16,1,1]` zones and a total problem domain `[-0.5,0.5]x[-0.5,0.5]x[-0.5,0.5]` and I create one particle at `t=0` at `x,y,z=-0.25,0,0` moving at `v=1` in the `+x` direction. I run on 2 MPI ranks until `tf=1` and this is my output (the `[%i]` in the output is `Globals::my_rank` i.e. corresponds to different meshblocks):
```
[0] push particle! xyz = -2.500000e-01 0.000000e+00 0.000000e+00
                    xyz = -2.187500e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -2.187500e-01 0.000000e+00 0.000000e+00
                    xyz = -1.875000e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -1.875000e-01 0.000000e+00 0.000000e+00
                    xyz = -1.562500e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -1.562500e-01 0.000000e+00 0.000000e+00
                    xyz = -1.250000e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -1.250000e-01 0.000000e+00 0.000000e+00
                    xyz = -9.375000e-02 0.000000e+00 0.000000e+00
[0] push particle! xyz = -9.375000e-02 0.000000e+00 0.000000e+00
                    xyz = -6.250000e-02 0.000000e+00 0.000000e+00
[0] push particle! xyz = -6.250000e-02 0.000000e+00 0.000000e+00
                    xyz = -3.125000e-02 0.000000e+00 0.000000e+00
[0] push particle! xyz = -3.125000e-02 0.000000e+00 0.000000e+00
                    xyz = 0.000000e+00 0.000000e+00 0.000000e+00
[1] push particle! xyz = 0.000000e+00 0.000000e+00 0.000000e+00
                    xyz = 3.125000e-02 0.000000e+00 0.000000e+00
[1] push particle! xyz = 3.125000e-02 0.000000e+00 0.000000e+00
                    xyz = 6.250000e-02 0.000000e+00 0.000000e+00
[1] push particle! xyz = 6.250000e-02 0.000000e+00 0.000000e+00
                    xyz = 9.375000e-02 0.000000e+00 0.000000e+00
[1] push particle! xyz = 9.375000e-02 0.000000e+00 0.000000e+00
                    xyz = 1.250000e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 1.250000e-01 0.000000e+00 0.000000e+00
                    xyz = 1.562500e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 1.562500e-01 0.000000e+00 0.000000e+00
                    xyz = 1.875000e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 1.875000e-01 0.000000e+00 0.000000e+00
                    xyz = 2.187500e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 2.187500e-01 0.000000e+00 0.000000e+00
                    xyz = 2.500000e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 2.500000e-01 0.000000e+00 0.000000e+00
                    xyz = 2.812500e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 2.812500e-01 0.000000e+00 0.000000e+00
                    xyz = 3.125000e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 3.125000e-01 0.000000e+00 0.000000e+00
                    xyz = 3.437500e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 3.437500e-01 0.000000e+00 0.000000e+00
                    xyz = 3.750000e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 3.750000e-01 0.000000e+00 0.000000e+00
                    xyz = 4.062500e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 4.062500e-01 0.000000e+00 0.000000e+00
                    xyz = 4.375000e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 4.375000e-01 0.000000e+00 0.000000e+00
                    xyz = 4.687500e-01 0.000000e+00 0.000000e+00
[1] push particle! xyz = 4.687500e-01 0.000000e+00 0.000000e+00
                    xyz = 5.000000e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -5.000000e-01 0.000000e+00 0.000000e+00
                    xyz = -4.687500e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -4.687500e-01 0.000000e+00 0.000000e+00
                    xyz = -4.375000e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -4.375000e-01 0.000000e+00 0.000000e+00
                    xyz = -4.062500e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -4.062500e-01 0.000000e+00 0.000000e+00
                    xyz = -3.750000e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -3.750000e-01 0.000000e+00 0.000000e+00
                    xyz = -3.437500e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -3.437500e-01 0.000000e+00 0.000000e+00
                    xyz = -3.125000e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -3.125000e-01 0.000000e+00 0.000000e+00
                    xyz = -2.812500e-01 0.000000e+00 0.000000e+00
[0] push particle! xyz = -2.812500e-01 0.000000e+00 0.000000e+00
                    xyz = -2.500000e-01 0.000000e+00 0.000000e+00
 ```
 so it looks like this particle is moving correctly across boundaries and meshblocks. I'll look at your example but I wanted to report this datapoint now.

File: example/particle_leapfrog/particle_leapfrog.cpp (hunk 1 above)
Comment (pgrete) [reply example/particle_leapfrog/particle_leapfrog.cpp]: Correct. Not needed and removed.

**Comment by pgrete** [comment]
@brryan the current version seems to work with one block per rank, but when I try to run with more than one the code hangs, e.g.,
`./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog` is fine and the particles end up at their initial positions, but
`./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog parthenon/meshblock/nx1=8` hangs indefinitely it cycle 0 and
`mpirun -np 2 ./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog parthenon/meshblock/nx1=8` works again.

Independently, trying to add <1 particles currently fails, e.g.,
```
$ mpirun -np 4 ./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog parthenon/meshblock/nx1=8  parthenon/meshblock/nx2=8
...
my particles
Rank 2 added particle 0 to block 2.
Rank 2 added particle 2 to block 2.
Rank 1 added particle 1 to block 1.
    	id                  Integer
    	vx                  Real
    	vy                  Real
    	vz                  Real

### PARTHENON ERROR
  Condition:   num_to_add > 0
  Message:     Attempting to add fewer than 1 new particles!
  File:        ../src/interface/swarm.cpp
  Line number: 242
Attempting to add fewer than 1 new particles![jever:181871] *** Process received signal ***
```

but I think this should work in principle (or is there a reason that blocks always need to have particles?)

**Comment by brryan** [comment]
@pgrete Ah OK I see. There are parts of the `Swarm` infrastructure that fall down if you try to create zero particles (allocating zero-size arrays, etc.). So I added a requirement that `AddEmptyParticles` only be called with >0 particles. I should lift this restriction and instead return some appropriately uninitialized `ParArrayND`. I'll try to do this now.

**Comment by brryan** [comment]
@pgrete OK there was another issue -- `InitializeCommunicationMesh` needs to be a loop over all blocks before `Send`/`Receive` calls, but I was calling this inside of the asynchronous region loop over `MeshBlock`s. The hanging you saw was due to some `Send` buffer flags being overwritten by repeated calls to `InitializeCommunicationMesh` which resets MPI flags. `InitializeCommunicationMesh` is now its own synchronous region over all blocks. Thanks for noticing this.

I fixed that and the add zero issue and pushed my changes to this branch, so all of your

`./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog`
`./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog parthenon/meshblock/nx1=8`
`mpirun -np 2 ./example/particle_leapfrog/particle-leapfrog -i ../example/particle_leapfrog/parthinput.particle_leapfrog parthenon/meshblock/nx1=8`

calls work for me.

**Review by brryan — APPROVED** [review]
By the way I approve :) This new example will be really nice to have going forward, both for the manual particle injection and the different pattern that `Send`/`Receive` calls take.

**Review by Yurlungur — APPROVED** [review]

**Comment by brryan** [comment]
@pgrete OK I think I addressed your comments from #497. Specifically,

- `ProblemGenerator` now requires 1D
- I split the example into two packages, `advection_package` and `particles_package`
- I added a comment to that spaghetti logic integration used to distribute the initial tracer population across `MeshBlock`s without MPI calls
- I divided the initial sine wave by `x_max_mesh - x_min_mesh` to enforce one wave period over the mesh.

Does that cover your concerns with this PR?

**Comment by pgrete** [comment]
Thanks. I'll squash and merge and then update the `brryan/particles_bounds` to develop and continue there.
